### PR TITLE
feat: Schema validator

### PR DESCRIPTION
**Pull request by klcodanr**
- Define a schema for Ingestion Requests
- Add a SchemaValidator class for validating object schema compliance

### CODE REVIEW

File: src/schema-validator.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+export class SchemaValidator {
+  #loaded = false;
+
+  #schemas = {};
+
+  async #getSchema(name) {
+    if (!this.#loaded) {
+      const schemasDir = join(
+        dirname(fileURLToPath(import.meta.url)),
```
Comment (mfrisbey) [inline src/schema-validator.js:25]: I'm unfamiliar with `import` in this case - what is this doing?

File: src/schema-validator.js
Comment (mfrisbey) [inline src/schema-validator.js:1]: My general feedback here would be to somehow describe or document the schema files work. i.e. if I want to add a new schema, how would I do that? What should the content of the schema file look like?

**Review by mfrisbey — APPROVED** [review]
I really like the concept of the schema validators, and would suggest that all of our services use a schema for their validation, instead of manually checking inputs.

File: src/schema-validator.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        'schemas',
+      );
+      const schemas = await readdir(schemasDir);
+      await Promise.all(
+        schemas
+          .filter((f) => f.endsWith('.json'))
+          .map(async (fn) => {
+            const schemaName = fn.slice(0, -5);
```
Comment (jdelbick) [inline src/schema-validator.js:33]: are you removing the `.json` here? 
you could also use: `fn.replace(/\.[^/.]+$/, '')`

File: src/schema-validator.js (hunk 1 above)
Comment (klcodanr) [reply src/schema-validator.js:25]: It's finding the URL of the current script on the system. This allows us to read the schemas via readFile from a relative directory in this module without having to use the JSON import syntax which isn't supported by eslint.

File: src/schema-validator.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    return this.#schemas[name];
+  }
+
+  /**
+   *
+   * @param {any} ingestionRequest
+   * @param {Array<string>} additionalRequiredData
+   */
+  async validateIngestionRequest(ingestionRequest, additionalRequiredData) {
+    const schema = await this.#getSchema('ingestion-request');
```
Comment (jdelbick) [inline src/schema-validator.js:49]: nitpick: make constant variable at the top of the file so its easier if we change it in the future

File: src/schema-validator.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    if (result.errors?.length > 0) {
+      throw new Error(
+        `Failed to validate schema, errors: \n- ${result.errors
+          .map((e) => e.message)
+          .join('\n- ')}`,
+      );
+    }
+    if (additionalRequiredData) {
+      const missing = additionalRequiredData.filter(
+        (k) => !(k in ingestionRequest.data),
```
Comment (jdelbick) [inline src/schema-validator.js:62]: if at all possible, avoid singular letter variable names so it is more descriptive for other contributers

**Review by jdelbick — APPROVED** [review]
LGTM, do we want to link this document: https://wiki.corp.adobe.com/display/WEM/Ingestor+API+Contract?

**Pull request merged**